src/order/irreducible.lean

```diff
@@ -0,0 +1,229 @@
+/-
+Copyright (c) 2023 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.finset.lattice
+import data.fintype.card
+
+/-!
+# Irreducible and prime elements in an order
+
+This file defines irreducible and prime elements in an order and shows that in a well-founded
+lattice every element decomposes as a supremum of irreducible elements.
+
+An element is sup-irreducible (resp. inf-irreducible) if it isn't `⊥` and can't be written as the
+supremum of any strictly smaller elements. An element is sup-prime (resp. inf-prime) if it isn't `⊥`
+and is greater than the supremum of any two elements less than it.
+
+Primality implies irreducibility in general. The converse only holds in distributive lattices.
+Both hold for all (non-minimal) elements in a linear order.
+
+## Main declarations
+
+* `sup_irred a`: Sup-irreducibility, `a` isn't minimal and `a = b ⊔ c → a = b ∨ a = c`
+* `inf_irred a`: Inf-irreducibility, `a` isn't maximal and `a = b ⊓ c → a = b ∨ a = c`
+* `sup_prime a`: Sup-primality, `a` isn't minimal and `a ≤ b ⊔ c → a ≤ b ∨ a ≤ c`
+* `inf_irred a`: Inf-primality, `a` isn't maximal and `a ≥ b ⊓ c → a ≥ b ∨ a ≥ c`
+* `exists_sup_irred_decomposition`/`exists_inf_irred_decomposition`: Decomposition into irreducibles
+  in a well-founded semilattice.
+-/
+
+open finset order_dual
+
+variables {ι α : Type*}
+
+/-! ### Irreducible and prime elements -/
+
+section semilattice_sup
+variables [semilattice_sup α] {a b c : α}
+
+/-- A sup-irreducible element is a non-bottom element which isn't the supremum of anything smaller.
+-/
+def sup_irred (a : α) : Prop := ¬ is_min a ∧ ∀ ⦃b c⦄, b ⊔ c = a → b = a ∨ c = a
```

Comment:
This definition has the disadvantage of not working for `finset α` without `decidable_eq α`. Do you think that's worth a TODO?

Reply:
I doubt it really matters. Even though I will indeed set `α := finset α` in later code, this will be done in the context of category theory, where we don't have the luxury of restricting to types with decidable equality.